### PR TITLE
Support data annotations PO localization

### PR DESCRIPTION
**Pull request by hishamco**
Address #4664

### CODE REVIEW

**Comment by hishamco** [comment]
I need to add some unit tests

**Comment by Skrypt** [comment]
Please see https://github.com/OrchardCMS/OrchardCore/issues/2974
and
https://github.com/OrchardCMS/OrchardCore/pull/3396

I'm not sure why this issue is coming back as it was fixed.

**Comment by hishamco** [comment]
Thanks Jasmin, I will check them ..

File: src/OrchardCore/OrchardCore.Localization.Core/ServiceCollectionExtensions.cs
Comment (jtkech) [inline src/OrchardCore/OrchardCore.Localization.Core/ServiceCollectionExtensions.cs]: i don't think it's a good pattern, better to use a transient option setup service (that will inject what it needs) registered through a module or our `OC.Mvc.Core startup`, but anyway after the mvc one, and then just use it to override the options set by mvc.

Comment (hishamco) [reply src/OrchardCore/OrchardCore.Localization.Core/ServiceCollectionExtensions.cs]: There's an extension for `IMvcBuilder` I think, but it require `.AddMvc()`

**Comment by hishamco** [comment]
@Skrypt if you look closely to #3396 I don't see anything related to data annotations at all seems it fixes another localizer bug

**Comment by Skrypt** [comment]
I could re-read the issue messages that @sebastienros made but I believe that he closed that issue after merging that PR so it must have been related. It was not introducing the services.AddDataAnnotationsPortableObjectLocalization(); in the DI pipeline though. 2 different things.

**Comment by hishamco** [comment]
AFAIK to make the data annotations works with custom localizer you should set `MvcDataAnnotationsLocalizationOptions.DataAnnotationLocalizerProvider` I already faced same issue with my `JsonLocalizer` that I started few years ago. Anyhow I will debug and fix the issue

**Comment by Skrypt** [comment]
Oh but what I meant is that if we want to enable this by default it should be decided first. I don't know why we would want this to be enabled by default in OC for example. So if we have this, then it should be configurable from a Feature.

**Comment by hishamco** [comment]
> So if we have this, then it should be configurable from a Feature.

That's what I'm thinking about before I begin this PR, ok let me finalize the unit tests, also we can ping @sebastienros to discuss this and the possibility to have it as a feature

**Comment by hishamco** [comment]
I found the bug and fix it, I will update the PR soon, but still surprising how the data annotation localized without injecting `AddDataLocalizations()` or is there anything I don't about

/cc @jtkech

**Comment by jtkech** [comment]
We call the following in the startup of `OC.Mvc.Core`.

            var builder = services.AddMvc(options =>
                ...
            });
            ...
            ...
            ...
            // Adding localization
            builder.AddViewLocalization();
            builder.AddDataAnnotationsLocalization();

**Comment by hishamco** [comment]
I see, but how the data annotation will aware about the `PortableObjectStringLocalizer`?

**Comment by jtkech** [comment]
I don't remember the details ;)
Maybe just because we are overriding the default string / html localizers.

**Comment by hishamco** [comment]
> Maybe just because we are overriding the default string / html localizers.

I don't think so, I already implemented a `JsonStringLocalizer` years ago I face the same exact issue, anyway .. thanks for your time and waiting for Seb comments

**Comment by jtkech** [comment]
Yes, not sure ;)  I would need to look at it more in depth. But after a quick look in the aspnetcore repo, seems that the default validator providers are using the following code that uses the registered `StringLocalizerFactory` that, as i remember, we are overriding

            if (_options.Value.DataAnnotationLocalizerProvider != null && _stringLocalizerFactory != null)
            {
                // This will pass first non-null type (either containerType or modelType) to delegate.
                // Pass the root model type(container type) if it is non null, else pass the model type.
                stringLocalizer = _options.Value.DataAnnotationLocalizerProvider(
                    context.ModelMetadata.ContainerType ?? context.ModelMetadata.ModelType,
                    _stringLocalizerFactory);
            }

File: src/OrchardCore/OrchardCore.Localization.Core/ServiceCollectionExtensions.cs
Comment (sebastienros) [reply src/OrchardCore/OrchardCore.Localization.Core/ServiceCollectionExtensions.cs]: Agreed with @jtkech, don't build the service collection, ever.

**Comment by hishamco** [comment]
@sebastienros before I cleanup the PR, regarding the `AddDataAnnotationsPortableObjectLocalization()`

- Shall we add it to localization module, so it can work outside OC too
- Shall we add it as feature

**Comment by hishamco** [comment]
⌚️ 📆

**Comment by agriffard** [comment]
Can you please solve the conflict @hishamco ?

**Comment by Skrypt** [comment]
This issue is an annoying recurring one. Please fix soon.

**Comment by hishamco** [comment]
Sure, I will resolve the conflict soon

**Comment by hishamco** [comment]
@Skrypt your final review

**Comment by Skrypt** [comment]
Have you tried with ErrorMessages on Data annotations. Lately I've merged latest dev and the login error messages started to display back in english all the time.

**Comment by hishamco** [comment]
Nope, I trust the unit tests :) but I will try this

**Comment by hishamco** [comment]
@Skrypt it's working fine with my branch, have a look

Login

**Comment by hishamco** [comment]
Could you try my branch before merge? Meanwhile I will resolve the conflict

**Comment by Skrypt** [comment]
I don't see any error validation messages on your screenshot of your login page.

**Comment by hishamco** [comment]
OMG seems I'm sleepy, I thought you are asked for the labels not the error messages 😊

I will check ASAP

**Comment by hishamco** [comment]
There's a bug, the error messages are not localized, I will check it ...

**Comment by hishamco** [comment]
> Have you tried with ErrorMessages on Data annotations. Lately I've merged latest dev and the login error messages started to display back in english all the time.

@Skrypt backing to your question. From where the default messages for require and display attributes are come from - before my PR -

**Comment by hishamco** [comment]
@Skrypt waiting for you so I can go further and close both bugs

**Comment by Skrypt** [comment]
`The {0} field is required. `

I know I tried to find this in the .po files and I think it was missing. I need to double check.

**Comment by Skrypt** [comment]
Capture d’écran, le 2019-12-14 à 14 10 35

**Comment by hishamco** [comment]
Seems there's a conflict here, I think I will split the PR into two, one to solve the inner class issue and the second one for data annotations

**Comment by Skrypt** [comment]
Yeah but the issue doesn't seem to be related to the fact that the error message is not set.

**Comment by hishamco** [comment]
May be, but I'm asking how can I get a localized version if the error messages are not set?!!

**Comment by hishamco** [comment]
I just create another PR to fix the original bug #4664, coz the bug is not related to the data annotation directly

I think it's fine to merge the other PR, and I will focus here with the data annotations

**Comment by Skrypt** [comment]
So the error message is the default one sent by ASP.NET Core from DataAnnotations. 

`The {0} field is required.`

But as you can see in my example I've set it explicitly in the ViewModel so it should try to get it's translation.

**Comment by hishamco** [comment]
> The {0} field is required.

Seems you 're right

> But as you can see in my example I've set it explicitly in the ViewModel so it should try to get it's translation.

I know this one :)

You know why Am I ask for this? I need to know the context for the translated message, this could be discover from `{culture}.po`, I'm still working on it along with other PR

**Comment by hishamco** [comment]
@Skrypt did you add a `*.po` file that contains the error messages and the French translations? Seems it should work if I'm not wrong, I will send an updated screenshot

**Comment by Skrypt** [comment]
Yes I did that without specifying any controller or view on the translation and it did not work.

**Comment by hishamco** [comment]
Strange!! I will show you mine ASAP with screenshot

**Comment by hishamco** [comment]
**Arabic Version**

```
msgid "The UserName field is required."
msgstr "حقل إسم المستخدم مطلوب."

msgid "The Password field is required."
msgstr "حقل كلمة المرور مطلوبة."

#. msgid "The {0} field is required."
#. msgstr "حقل ال {0} مطلوب."
```

Login-ar-YE

**French Version**

Similarly with

```
msgid "The UserName field is required."
msgstr "Le nom d'utilisateur est requis."

msgid "The Password field is required."
msgstr "Le champ Mot de passe est requis."
``` 

Login-fr

**Comment by hishamco** [comment]
FYI I just decorate the `LoginViewModel` as the following:

```csharp
    public class LoginViewModel
    {
        [Required(ErrorMessage = "The UserName field is required.")]
        public string UserName { get; set; }

        [Required(ErrorMessage = "The Password field is required.")]
        [DataType(DataType.Password)]
        public string Password { get; set; }

        public bool RememberMe { get; set; }
    }
```

but you could specify the default error message in case you don't want to modify the models

**Comment by hishamco** [comment]
> but you could specify the default error message in case you don't want to modify the models

I realize the default message aren't localized, when I debug it doesn't hit `LocalizedString this[string name]` nor `LocalizedString this[string name, params object[] arguments]`

**Comment by hishamco** [comment]
Frankly what main I did in this PR, is trigger the data annotation localization when localization module is enabled, and this is the proper place that it should be IMHO

I'm still investigating why the default data annotations messages are not localized or we should override them

**Comment by hishamco** [comment]
AFAIK this is a common issue https://github.com/dotnet/corefx/issues/25128, but we can fix it by introducing a new validation adapter, I'm trying to do it now and I will keep you on touch ...

**Comment by hishamco** [comment]
I tried to introduce a `LocalizedValidationAttributeAdapterProvider` I did a good work there, and I was able to localize the data annotation messages, but something I don't know about shows the default messages on the client :(

It takes long time to investigate and look into the MVC data annotations source, after that I decide to go with another approach by introducing a `LocalizedValidationMetadataProvider` which is works very nice :)

To summarize what I did or what's new in the PR, I can say:

- By adding enabling localization module the `AddDataAnnotationsPortableObjectLocalization()` will be called, which is responsible to 

- Register the `LocalizedDataAnnotationsMvcOptions` into the DI, which is under the hood

- Adds `LocalizedValidationMetadataProvider` to `MvcOptions.ModelMetadataDetailsProviders`

so what I did will enable us to:

1. Localize the default error messages which is not supported by default on ASP.NET Core :(
```csharp
[Required]
public property string UserName { get; set; }
```

The localization will be done in two ways:

1.1. The localizer will look for the `FormatErrorMessage()` in our example will be `The field UserName is required.` if it found, the localized string will be shown, otherwise

1.2. The localizer will look for the default error message that defines in the data annotations attributes themselves in our example will be `The field {0} is required.` 

2. Localize the data annotations error messages which supplied by the devs
```csharp
[Required(ErrorMessage: "The field UserName is required.")]
public property string UserName { get; set; }
```

IMHO this is cool because either I supply the error message on the data annotations or the localizer will look for the error message with and without params

@Skrypt could you please try my branch if you have time

/cc @sebastienros @jtkech your thoughts is very important to me

**Comment by jtkech** [comment]
Okay, i will take a look as i have time, so many things ;)

**Comment by hishamco** [comment]
What is the status of this item?

/cc @sebastienros @Skrypt

**Comment by hishamco** [comment]
@agriffard I tried my code many times, it works like a charm ;)

DataAnnotations Localization

Be sure you added the following files in the `Localization` folder to make the demo works:

**fr.po**
```
#. msgid "The UserName field is required."
#. msgstr "Le nom d'utilisateur est requis."

msgid "The Password field is required."
msgstr "Le champ Mot de passe est requis."

msgid "The {0} field is required."
msgstr "Le {0} est requis."

msgid "The value '{0}' is invalid."
msgstr "La valeur '{0}' n'est pas valide."
```
**ar.po**
```
msgid "The UserName field is required."
msgstr "حقل إسم المستخدم مطلوب."

#. msgid "The Password field is required."
#. msgstr "حقل كلمة المرور مطلوبة."

msgid "The {0} field is required."
msgstr "حقل ال {0} مطلوب."

msgid "The value '' is invalid."
msgstr "القيمة غير صحيحة."
```

**Comment by hishamco** [comment]
@sebastienros I joined the triage yesterday, nothing to decide about this PR, IMHO  it's fine to provide the default error message in `en.po` so for other cultures we can use Crowdin to translate them. If a user intend to customize the error messages in the data annotations attributes themselves, we can extend the `PoExtractor` to do the job for us. 

All the feedback are welcome .. thanks

**Comment by hishamco** [comment]
@agriffard please let me know once your merge this #5793 to add some unit tests if it's possible to ensure that data annotations localizations work for both default and custom messages

File: src/OrchardCore/OrchardCore.Localization.Core/Extensions/CultureDictionaryExtensions.cs
Comment (hishamco) [inline src/OrchardCore/OrchardCore.Localization.Core/Extensions/CultureDictionaryExtensions.cs]: May be it's bad idea to ignore the `context` but this is the way to come up with to get the data annotations message if there is, because the context all the time will be `OrchardCore.Localization.DataAnnotations.LocalizedDataAnnotationsMvcOptions`

@jtkech @sebastienros if there's a better way, please shoot it to me

File: src/OrchardCore/OrchardCore.Localization.Core/DataAnnotations/LocalizedDataAnnotationsMvcOptions.cs
Comment (jtkech) [inline src/OrchardCore/OrchardCore.Localization.Core/DataAnnotations/LocalizedDataAnnotationsMvcOptions.cs]: When executed in a tenant context there is already an ambient scope service provider that you can access through an `IHttpContextAccessor` that you could inject in the ctor, or directly by using `ShellScope.Services`.

But but, because `IStringLocalizerFactory` seems to be a tenant singleton, you can inject it directly in the ctor and then use it in the `Configure()` method without having to create a scope.

File: src/OrchardCore/OrchardCore.Localization.Core/PortableObject/PortableObjectStringLocalizer.cs
Comment (jtkech) [inline src/OrchardCore/OrchardCore.Localization.Core/PortableObject/PortableObjectStringLocalizer.cs]: Would need more time to understand what happens here. Not sure of the name, maybe `GetTranslationWithoutContext()`.

Did you tried it? Seems to do quite the same as `dictionary[key]` => `dictionary[key, null]`, the only difference i see is that you don't take into account the current `count` and you lookup in all `dictionary.Translations`, `Translations` being a `Dictionary<string, string[]>`. Does it mean that here we can't have a valid `count`?

If so maybe `GetTranslationWithoutCount()`, or maybe `GetTranslationWithoutPluralRule()`, not sure ...

File: src/OrchardCore/OrchardCore.Localization.Core/Extensions/CultureDictionaryExtensions.cs
Comment (jtkech) [reply src/OrchardCore/OrchardCore.Localization.Core/Extensions/CultureDictionaryExtensions.cs]: I made a comment below where you use it

File: src/OrchardCore/OrchardCore.Localization.Core/PortableObject/PortableObjectStringLocalizer.cs
Comment (hishamco) [reply src/OrchardCore/OrchardCore.Localization.Core/PortableObject/PortableObjectStringLocalizer.cs]: > Seems to do quite the same as dictionary[key] => dictionary[key, null]

Nope, coz If you look closely into the extension method, I'm looking for the key such as `RequiredAttributeDefaultMessage` in all the translations and the result something like `OrchardCore.DataAnnotations|RequiredAttributeDefaultMessage`

So `dictionary[key, null]` will look for key named `RequiredAttributeDefaultMessage` and not `OrchardCore.DataAnnotations|RequiredAttributeDefaultMessage`

File: src/OrchardCore/OrchardCore.Localization.Core/DataAnnotations/LocalizedDataAnnotationsMvcOptions.cs
Comment (hishamco) [reply src/OrchardCore/OrchardCore.Localization.Core/DataAnnotations/LocalizedDataAnnotationsMvcOptions.cs]: Let me try `ShellScope.Services`

> But but, because IStringLocalizerFactory seems to be a tenant singleton, you can inject it directly in the ctor and then use it in the Configure() method without having to create a scope.

Which `Configure` method do you mean?

Comment (jtkech) [reply src/OrchardCore/OrchardCore.Localization.Core/DataAnnotations/LocalizedDataAnnotationsMvcOptions.cs]: This one.

I meant inject `IStringLocalizerFactory` directly in your ctor

And then use it in your `public void Configure(MvcOptions options)`

Comment (hishamco) [reply src/OrchardCore/OrchardCore.Localization.Core/DataAnnotations/LocalizedDataAnnotationsMvcOptions.cs]: Got it, let me try

File: src/OrchardCore/OrchardCore.Localization.Core/PortableObject/PortableObjectStringLocalizer.cs
Comment (jtkech) [reply src/OrchardCore/OrchardCore.Localization.Core/PortableObject/PortableObjectStringLocalizer.cs]: Okay, would need to focus on it more, but i trust you ;)

Sorry, i'm a little lost here, what would be the value of `count` here?

How the `DataAnnotationsDefaultErrorMessages` that you defined in another PR will be used?

Comment (hishamco) [reply src/OrchardCore/OrchardCore.Localization.Core/PortableObject/PortableObjectStringLocalizer.cs]: `DataAnnotationsDefaultErrorMessages` this just for the `PoExtractor` to extract the localized strings nothing more ;)

File: src/OrchardCore/OrchardCore.Localization.Core/DataAnnotations/LocalizedDataAnnotationsMvcOptions.cs
Comment (hishamco) [reply src/OrchardCore/OrchardCore.Localization.Core/DataAnnotations/LocalizedDataAnnotationsMvcOptions.cs]: I update the PR, I'm not sure why I didn't inject `IStringLocalizerFactory` from the beginning ;)

File: src/OrchardCore/OrchardCore.Localization.Core/PortableObject/PortableObjectStringLocalizer.cs
Comment (hishamco) [reply src/OrchardCore/OrchardCore.Localization.Core/PortableObject/PortableObjectStringLocalizer.cs]: > Sorry, i'm a little lost here, what would be the value of count here?

The `count` is omitted

**Comment by hishamco** [comment]
This all the changes to make this happen, there's nothing to do here, except waiting for you final review ...

**Comment by rjpowers10** [comment]
@hishamco what is the status of this PR? I'm interested in translating the default "The {0} field is required." and I think this PR lets me do that.

**Comment by hishamco** [comment]
@rjpowers10  It's ready but we postponded after @agriffard introduced aworkaround. Believe it or not I'm just thinking about this PR two hours ago ;)

If you can test it I will be thankful, also I will try to double check later

**Comment by rjpowers10** [comment]
> @rjpowers10 It's ready but we postponded after @agriffard introduced aworkaround. Believe it or not I'm just thinking about this PR two hours ago ;)
> 
> If you can test it I will be thankful, also I will try to double check later

@hishamco I don't see that workaround, but it might just be because the conversation is pretty long. Do you have a link to the workaround?

**Comment by hishamco** [comment]
It's `IValidator` ro something similar that validate then use the default localizer. If you can test this PR and verify that's it works as expect I can push the team to merge this PR if it doesn't break anything

**Comment by dev-rlipa** [comment]
@hishamco Earlier today, I set up an example where localization of the error message for the `Required `attribute worked. However, I did run into a problem when using the `Display` attribute. 

I downloaded the data-annotation-localization branch and tested with `SmtpSettings.DefaultSender`, which has a `Required `attribute.
```
    public class SmtpSettings : IValidatableObject
    {
        /// <summary>
        /// Gets or sets the default sender mail.
        /// </summary>
        [Required(AllowEmptyStrings = false), EmailAddress]
        public string DefaultSender { get; set; }  
    }
```

I added an OrchardCore.Cms.Web/App_Data/Sites/Default/Localization/en-US.po file with these contents:
```
msgid "The {0} field is required."
msgstr "Translated message: {0} is required."
```

In the admin dashboard on the Configuration > Settings  > Smtp page, I submitted the form without specifying a sender email address, and the localized error message displayed, which was correct.

image

Then, I added a `Display `attribute to `SmtpSettings.DefaultSender` and specified a display name.

```
    public class SmtpSettings : IValidatableObject
    {
        /// <summary>
        /// Gets or sets the default sender mail.
        /// </summary>
        [Required(AllowEmptyStrings = false), EmailAddress]
        [Display(Name = "Default Sender")]
        public string DefaultSender { get; set; }  
       ...
    }
```

I modified the OrchardCore.Cms.Web/App_Data/Sites/Default/Localization/en-US.po file to provide a translation for the display name, so it looked like this:

```
msgid "The {0} field is required."
msgstr "Translated message: {0} is required."

msgid "Default Sender"
msgstr "Translated Default Sender"
```

I also modified SmtpSettings.Edit.cshtml, so the label for the default sender was not specified directly in the view and would be determined by the `Display `attribute.

`<label asp-for="DefaultSender"></label>`

In the admin dashboard on the Configuration > Settings  > Smtp page, I submitted the form without specifying a sender email address. The localized error message displayed, but there were two problems:

1. The label displayed as the un-localized display name ("Default Sender") instead of the localized display name ("Translated Default Sender").
2. The {0} parameter in the error message was replaced with the property name (DefaultSender) instead of the localized display name ("Translated Default Sender").

image

I also tried removing the translation for the default required error message from the en-US.po file, so it looked like this:
```
msgid "Default Sender"
msgstr "Translated Default Sender"
```

However, the label was still un-localized and the property name displayed in the error message, instead of the localized display name.

image

For (1) above, I have been able to get the localization of the `Display `attribute to work at a more recent version of Orchard Core (version 1.0.0-rc2-16161), so I'm not sure if this is an issue related to the PR or if it is because the data-annotation-localization branch is behind dev.

For (2) above, I wonder if LocalizedValidationMetadataProvider.CreateValidationMetadata needs to take the `Display ` attribute into account when it parameterizes the error message. Currently, it looks like it's just using the property name (`context.Key.Name`) as the `argument`.

```
public void CreateValidationMetadata(ValidationMetadataProviderContext context)
        {
            foreach (var metadata in context.ValidationMetadata.ValidatorMetadata)
            {
                var attribute = metadata as ValidationAttribute;
                var argument = context.Key.Name;
                var errorMessageString = attribute != null && attribute.ErrorMessage == null && attribute.ErrorMessageResourceName == null
                    ? attribute.FormatErrorMessage(argument)
                    : attribute.ErrorMessage;

                // Localize the error message without params
                var localizedString = _stringLocalizer[errorMessageString];

                if (localizedString == errorMessageString)
                {
                    // Localize the error message with params
                    errorMessageString = errorMessageString.Replace(argument, "{0}");
                    localizedString = _stringLocalizer[errorMessageString, argument];
                }

                attribute.ErrorMessage = localizedString;
            }
        }
    }
```

**Comment by hishamco** [comment]
Thanks a lot @dev-rlipa AFAIK there's an issue with the display attribute from a long long time, but I'm not sure if it got fixed in ASP.NET Core

I may try this PR after rebase on `dev` branch or you can change the `CreateValidationMetadata()` to make the `Display` works as well, I'm busy with a dynamic data localization PR, hopefully I got a time to revisit this

Thanks again

File: src/OrchardCore/OrchardCore.Localization.Core/DataAnnotations/LocalizedValidationMetadataProvider.cs
Comment (dev-rlipa) [inline src/OrchardCore/OrchardCore.Localization.Core/DataAnnotations/LocalizedValidationMetadataProvider.cs:38]: I think this won't take the `Display` attribute into account when parameterizing the error message.

Since the `argument ` (`context.Key.Name`) is the property name here, the property name will display in the error message, even if the view model property has a `Display `attribute and there is a translation in the PO file.

In other words, if I add a `Display` attribute to a view model property like this

```
[Display(Name = "Default Sender")]
public string DefaultSender { get; set; }
```
and have entries in my en-US.po file like this

```
msgid "The {0} field is required."
msgstr "Translated message: {0} is required."

msgid "Default Sender"
msgstr "Translated Default Sender"
```

the localized error message will be "Translated message: DefaultSender is required."

However, I would expect the localized error message to take into account the localized display name, so it would be "Translated message: Translated Default Sender is required."

I think we can leave the error message unparameterized here.

```suggestion
                    localizedString = _stringLocalizer[errorMessageString];
```

It looks like [DataAnnotationsMetadataProvider.CreateDisplayMetadata](https://github.com/aspnet/Mvc/blob/master/src/Microsoft.AspNetCore.Mvc.DataAnnotations/DataAnnotationsMetadataProvider.cs#L151) ultimately parameterizes the error message and takes into account the localized display name.

image

If we make the suggested change above and add `builder.AddDataAnnotationsLocalization()` back into OrchardCore.Mvc.Core.Startup.cs (as suggested in a separate comment in this review), the localized error message correctly includes the localized display name as a parameter.

image

**Review by dev-rlipa — CHANGES_REQUESTED** [review]

**Comment by dev-rlipa** [comment]
@hishamco I have suggestions to fix issues (1) and (2), which we discussed above. Please see my code review comments.

**Comment by hishamco** [comment]
Thanks a lot @dev-rlipa I have an Internet issue, hope to check your reviews ASAP

**Comment by hishamco** [comment]
@dev-rlipa just few updates on this PR, now you will be able to localize `DisplayAttribute`, but be sure that the message context in PO is coming from `ViewModel` not the `Model`. Regarding the normal error messages they should work as well either from the default error messages that are listed in `DataAnnotationsDefaultErrorMessages` or from a custom `ErrorMessage` in the data annotation attributes. I just tested all the cases and everything is fine, please double check and let me if something doesn't go well

Thanks

**Comment by hishamco** [comment]
@agriffard could you please check the Login, the message should be localized without implementing `IValidatableObject`, If everything goes well which what I expect, I think I will replace all the occurenances from your old PR #5692

**Comment by hishamco** [comment]
> However, I would expect the localized error message to take into account the localized display name, so it would be "Translated message: Translated Default Sender is required."

I think I miss to test this case ;)

**Comment by dev-rlipa** [comment]
Thank you for following up, @hishamco. I downloaded the recent changes and tested them out using my same example as above. Unfortunately, I am still observing the same issues with the `Display` attribute not localizing, both for the labels and the error messages.

image

To solve these issues, I think we'll need to make the changes recommended in the code review.

In `LocalizedValidationMetadataProvider.CreateValidationMetadata`, I think we should leave the error message un-parameterized like this

```
localizedString = _stringLocalizer[errorMessageString];
```

instead of this

```
localizedString = _stringLocalizer[errorMessageString, argument];
```

Additionally, I think we should add

`builder.AddDataAnnotationsLocalization();`

back into the `Startup.cs` file for OrchardCore.Mvc.Core.

With these above 2 changes, the `Display` attribute correctly localizes, both for the labels and the error messages.

image

**Comment by hishamco** [comment]
One more not I forgot to mention please make sure that the data annotations resources are updated coz I suffered at the beginning and nothing work, then I realised it's a po file issue, so I bring the updated version from Orchard Translations repo

I can provide a screenshot if it requires

**Comment by dev-rlipa** [comment]
@hishamco, that would be great if you could provide a screenshot, so I can double-check.

Note that I'm testing with my own en-US.po file, instead of one of the[ OrchardCore.Translations](OrchardCore.Translations) PO files. I placed my en-US.po file under OrchardCore.Cms.Web\App_Data\Sites\Default\Localization and the content of my en-US.po file looks like this:

```
msgid "The {0} field is required."
msgstr "Translated message: {0} is required."

msgid "Default Sender"
msgstr "Translated Default Sender"
```

I omitted the context (`msgctxt`) elements, since they are optional. For testing purposes, I also made changes to SmtpSettings.cs and SmtpSettings.Edit.cshtml, as described above.

**Comment by hishamco** [comment]
Sorry for late @dev-rlipa I run into an issue with the internet connection, I will try to post a screenshot, BTW, if you change your translations, they will be replaced coz there's a build task to copy all the translations

**Comment by hishamco** [comment]
Sorry for late, I moved to a new house, but unfortuently there's no internet connection yet, that's why I was not online last few weeks

Hope to create screenshot today ...

**Comment by rjpowers10** [comment]
Thanks, @hishamco! We are still very interested in this.

**Comment by hishamco** [comment]
Believe it or not my laptop is hungging for awhile after fetching the latest changes and trying to restore and run the app ;)

**Comment by hishamco** [comment]
The below screenshot show how to localization the data annotations (DisplayName included) without message context

Data Annotation Localization without Context

**Comment by hishamco** [comment]
This one with message context which is the default behavior

Data Annotation Localization with Context

/cc @dev-rlipa hope it's clear now

**Comment by hishamco** [comment]
FYI the only limitation that's obvious is `asp-for` with `DisplayNameAttribute` is not localization, OC used `@T[""]` in all the views, so we can look into this later if it's requested

**Comment by dev-rlipa** [comment]
Thank you for the helpful screenshots, @hishamco. They seem to confirm that our PO files were setup correctly.

I tested out the most recent version of this branch, and the parameterized error messages are now taking into account the display name. Unfortunately, though, I noticed that the display name is not getting localized in the parameterized error message.

image

If my PO file translates "Default Sender" to "Translated Default Sender," I would instead expect the parameterized error message to be "Translated message: **Translated Default Sender** is required."

Regarding

> FYI the only limitation that's obvious is `asp-for` with `DisplayNameAttribute` is not localization, OC used `@T[""]` in all the views, so we can look into this later if it's requested

we are interested in support for this scenario if possible. We are planning to use the display name attribute to generate the labels in our views, instead of `@T[""]`, so it will be important for the display name to get localized.

To summarize, there are 2 main issues that we're encountering:
(1) The parameters in the error messages are not getting localized when we use the display name attribute.
(2) The labels are not getting localized when we use `asp-for` with the the display name attribute.

As I mentioned in a [previous comment](https://github.com/OrchardCMS/OrchardCore/pull/4675#issuecomment-842350004), I think I have a fix for these issues.

> To solve these issues, I think we'll need to make the changes recommended in the code review.
> 
> In `LocalizedValidationMetadataProvider.CreateValidationMetadata`, I think we should leave the error message un-parameterized like this
> 
> ```
> localizedString = _stringLocalizer[errorMessageString];
> ```
> 
> instead of this
> 
> ```
> localizedString = _stringLocalizer[errorMessageString, argument];
> ```
> 
> Additionally, I think we should add
> 
> `builder.AddDataAnnotationsLocalization();`
> 
> back into the `Startup.cs` file for OrchardCore.Mvc.Core.
> 
> With these above 2 changes, the `Display` attribute correctly localizes, both for the labels and the error messages.
> 
> image

I tested out the changes above with the most recent version of this branch, and they seem to work.

@hishamco, could you try out the changes above and let me know what you think? I appreciate all your help looking into this.

**Comment by hishamco** [comment]
> Unfortunately, though, I noticed that the display name is not getting localized in the parameterized error message.

This is doable, but I suggest that you can provide the localized resource without params for now, something like

```
msgid "The Display Name field is required."
msgstr "Translated message: Display Name field is required."
```

**Comment by hishamco** [comment]
For now I need to make sure that guys confirm that the data localization works fine, then merge this PR. After that we could have a look to support `DisplayNameAttribute` in `asp-for`

**Comment by hishamco** [comment]
@Skrypt @agriffard could you please - if you have a time to - confirm that the PR is worked as supposed to be

**Comment by hishamco** [comment]
@dev-rlipa you could clone this branch and apply your fix and may I need to spend some time to test all the cases that I made previously

**Comment by hishamco** [comment]
The PR is ready to merge, there are two requested scenarios by the community that addressed in #9723

**Comment by hishamco** [comment]
@dev-rlipa your changes has beed added here

Thanks

**Comment by rjpowers10** [comment]
@hishamco Just a gentle reminder that we are still interested in this 😊

**Comment by hishamco** [comment]
> @hishamco Just a gentle reminder that we are still interested in this 😊

Me too ;) @sebastienros can we take an action about this one, It's around two years ;) @rjpowers10 @dev-rlipa confirms that it works as expected, I think we need someone from the core team - if it's possible - to confirm

**Comment by hishamco** [comment]
@sebastienros I think it's the time to consider this, it was too long since we start this ;)

@dev-rlipa @rjpowers10 please can you confirm again if this PR works as expected, while you are guys use it?

**Comment by dev-rlipa** [comment]
@hishamco and @rjpowers10 I tested out this PR again and confirmed that it works as expected.

**Comment by hishamco** [comment]
Thanks a lot @dev-rlipa for you efforts on this PR, @@sebastienros can we take an action on this PR

/cc @Skrypt @agriffard @jtkech

**Comment by Skrypt** [comment]
I think this is the PR we should all start putting efforts on. If everything else is fixed and stable for 1.2.0.

**Comment by sebastienros** [comment]
Test and merge. Ensure it's still usable without other reference to Orchard Core. Good for 1.2

**Comment by hishamco** [comment]
I will do a final test

> Ensure it's still usable without other reference to Orchard Core

Which reference you mean?

**Comment by Skrypt** [comment]
I need clear documentation on usage for this because I had issues using it. Also, not clear how to use `D["something"]`. I'm not even sure this PR contains the change I had done on my side.

**Comment by hishamco** [comment]
Tested and merged :rocket :rocket :rocket

**Comment by hishamco** [comment]
Thanks again @dev-rlipa @rjpowers10

**Pull request merged**
